File: Deets/ViewModels/ContactPreviewViewModel.swift
//
//  ContactPreviewViewModel.swift
//  Deets
//
//  ViewModel for contact preview and editing
//

import SwiftUI
import SwiftData
import Contacts
import Observation

@Observable
@MainActor
final class ContactPreviewViewModel {
    // MARK: - Published State

    /// Editable fields
    var fullName = ""
    var jobTitle = ""
    var company = ""
    var email = ""
    var phoneNumber = ""
    var website = ""
    var address = ""
    var notes = ""

    /// Validation states
    var isValidEmail = true
    var isValidPhone = true
    var isValidWebsite = true

    /// Loading and error states
    var isSaving = false
    var saveError: String?
    var showSuccessAlert = false

    /// Contact save status
    var savedToContacts = false

    // MARK: - Private Properties

    private let rawText: String
    private var modelContext: ModelContext?

    // MARK: - Dependencies

    private let hapticManager = HapticManager.shared

    // MARK: - Initialization

    init(scannedText: String) {
        self.rawText = scannedText
        parseScannedText(scannedText)
    }

    /// Set model context for saving
    func setModelContext(_ context: ModelContext) {
        self.modelContext = context
    }

    // MARK: - Parsing

    /// Parse scanned text into contact fields
    private func parseScannedText(_ text: String) {
        let lines = text.components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }

        // Simple parsing logic - can be enhanced with NLP later
        for (index, line) in lines.enumerated() {
            if line.contains("@") && email.isEmpty {
                email = line
            } else if line.hasPrefix("http") || line.hasPrefix("www") {
                website = line
            } else if line.matches(phonePattern) {
                phoneNumber = line
            } else if index == 0 && fullName.isEmpty {
                fullName = line
            } else if index == 1 && jobTitle.isEmpty {
                jobTitle = line
            } else if index == 2 && company.isEmpty {
                company = line
            }
        }

        validateAllFields()
    }

    // MARK: - Validation

    /// Phone number regex pattern
    private let phonePattern = #"^[\+]?[(]?[0-9]{1,4}[)]?[-\s\.]?[(]?[0-9]{1,4}[)]?[-\s\.]?[0-9]{1,9}$"#

    /// Validate email format
    func validateEmail() {
        guard !email.isEmpty else {
            isValidEmail = true
            return
        }
        let emailRegex = #"^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$"#
        isValidEmail = email.range(of: emailRegex, options: [.regularExpression, .caseInsensitive]) != nil
    }

    /// Validate phone format
    func validatePhone() {
        guard !phoneNumber.isEmpty else {
            isValidPhone = true
            return
        }
        isValidPhone = phoneNumber.range(of: phonePattern, options: .regularExpression) != nil
    }

    /// Validate website URL
    func validateWebsite() {
        guard !website.isEmpty else {
            isValidWebsite = true
            return
        }
        if let url = URL(string: website) {
            isValidWebsite = url.scheme != nil && url.host != nil
        } else {
            isValidWebsite = false
        }
    }

    /// Validate all fields
    func validateAllFields() {
        validateEmail()
        validatePhone()
        validateWebsite()
    }

    // MARK: - Computed Properties

    /// Whether the form has valid data to save
    var canSave: Bool {
        !fullName.isEmpty && isValidEmail && isValidPhone && isValidWebsite
    }

    /// Whether any field has been modified
    var hasChanges: Bool {
        !fullName.isEmpty || !email.isEmpty || !phoneNumber.isEmpty
    }

    // MARK: - Saving

    /// Save business card to SwiftData
    func saveToDatabase() async throws {
        guard let context = modelContext else {
            throw SaveError.noContext
        }

        guard canSave else {
            throw SaveError.invalidData
        }

        isSaving = true
        defer { isSaving = false }

        let card = BusinessCard(
            fullName: fullName.trimmingCharacters(in: .whitespaces),
            jobTitle: jobTitle.isEmpty ? nil : jobTitle.trimmingCharacters(in: .whitespaces),
            company: company.isEmpty ? nil : company.trimmingCharacters(in: .whitespaces),
            email: email.isEmpty ? nil : email.trimmingCharacters(in: .whitespaces),
            phoneNumber: phoneNumber.isEmpty ? nil : phoneNumber.trimmingCharacters(in: .whitespaces),
            website: website.isEmpty ? nil : website.trimmingCharacters(in: .whitespaces),
            address: address.isEmpty ? nil : address.trimmingCharacters(in: .whitespaces),
            notes: notes.isEmpty ? nil : notes.trimmingCharacters(in: .whitespaces),
            rawText: rawText,
            savedToContacts: savedToContacts
        )

        context.insert(card)

        do {
            try context.save()
            hapticManager.saved()
            showSuccessAlert = true
        } catch {
            throw SaveError.databaseError(error)
        }
    }

    /// Save contact to iOS Contacts app
    func saveToContacts() async throws {
        let store = CNContactStore()

        // Request authorization
        let status = CNContactStore.authorizationStatus(for: .contacts)

        switch status {
        case .notDetermined:
            let granted = try await store.requestAccess(for: .contacts)
            if !granted {
                throw SaveError.contactsPermissionDenied
            }
        case .denied, .restricted:
            throw SaveError.contactsPermissionDenied
        case .authorized:
            break
        @unknown default:
            throw SaveError.contactsPermissionDenied
        }

        // Create contact
        let contact = CNMutableContact()
        contact.givenName = fullName.components(separatedBy: " ").first ?? ""
        contact.familyName = fullName.components(separatedBy: " ").dropFirst().joined(separator: " ")

        if !jobTitle.isEmpty {
            contact.jobTitle = jobTitle
        }

        if !company.isEmpty {
            contact.organizationName = company
        }

        if !email.isEmpty {
            contact.emailAddresses = [CNLabeledValue(label: CNLabelWork, value: email as NSString)]
        }

        if !phoneNumber.isEmpty {
            let phone = CNPhoneNumber(stringValue: phoneNumber)
            contact.phoneNumbers = [CNLabeledValue(label: CNLabelWork, value: phone)]
        }

        if !website.isEmpty {
            contact.urlAddresses = [CNLabeledValue(label: CNLabelWork, value: website as NSString)]
        }

        if !address.isEmpty {
            let homeAddress = CNMutablePostalAddress()
            homeAddress.street = address
            contact.postalAddresses = [CNLabeledValue(label: CNLabelWork, value: homeAddress)]
        }

        if !notes.isEmpty {
            contact.note = notes
        }

        // Save to contacts
        let saveRequest = CNSaveRequest()
        saveRequest.add(contact, toContainerWithIdentifier: nil)

        try store.execute(saveRequest)
        savedToContacts = true
        hapticManager.saved()
    }

    /// Save both to database and contacts
    func saveBoth() async {
        do {
            try await saveToContacts()
            savedToContacts = true
            try await saveToDatabase()
        } catch {
            saveError = error.localizedDescription
            hapticManager.scanError()
        }
    }

    /// Save only to database
    func saveToDatabaseOnly() async {
        do {
            try await saveToDatabase()
        } catch {
            saveError = error.localizedDescription
            hapticManager.scanError()
        }
    }
}

// MARK: - Errors

enum SaveError: LocalizedError {
    case noContext
    case invalidData
    case databaseError(Error)
    case contactsPermissionDenied

    var errorDescription: String? {
        switch self {
        case .noContext:
            return "Database context not available"
        case .invalidData:
            return "Please fix validation errors before saving"
        case .databaseError(let error):
            return "Failed to save: \(error.localizedDescription)"
        case .contactsPermissionDenied:
            return "Contacts permission is required to save to your contacts. Please enable it in Settings."
        }
    }
}

// MARK: - String Extension

private extension String {
    func matches(_ pattern: String) -> Bool {
        range(of: pattern, options: .regularExpression) != nil
    }
}

File: Android/app/src/main/kotlin/com/deets/presentation/preview/ContactPreviewViewModel.kt
package com.deets.presentation.preview

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.deets.domain.model.BusinessCard
import com.deets.domain.model.ParsedContact
import com.deets.domain.repository.BusinessCardRepository
import com.deets.services.ContactsService
import com.deets.util.PermissionManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.Date
import javax.inject.Inject

/**
 * ViewModel for Contact Preview screen
 * Mirrors iOS ContactPreviewViewModel
 */
@HiltViewModel
class ContactPreviewViewModel @Inject constructor(
    private val repository: BusinessCardRepository,
    private val contactsService: ContactsService,
    private val permissionManager: PermissionManager
) : ViewModel() {

    private val _uiState = MutableStateFlow<PreviewUiState>(PreviewUiState.Editing)
    val uiState: StateFlow<PreviewUiState> = _uiState.asStateFlow()

    private val _parsedContact = MutableStateFlow<ParsedContact?>(null)
    val parsedContact: StateFlow<ParsedContact?> = _parsedContact.asStateFlow()

    // Editable fields
    private val _fullName = MutableStateFlow("")
    val fullName: StateFlow<String> = _fullName.asStateFlow()

    private val _jobTitle = MutableStateFlow("")
    val jobTitle: StateFlow<String> = _jobTitle.asStateFlow()

    private val _company = MutableStateFlow("")
    val company: StateFlow<String> = _company.asStateFlow()

    private val _email = MutableStateFlow("")
    val email: StateFlow<String> = _email.asStateFlow()

    private val _phoneNumber = MutableStateFlow("")
    val phoneNumber: StateFlow<String> = _phoneNumber.asStateFlow()

    private val _website = MutableStateFlow("")
    val website: StateFlow<String> = _website.asStateFlow()

    private val _notes = MutableStateFlow("")
    val notes: StateFlow<String> = _notes.asStateFlow()

    /**
     * Initialize with parsed contact
     */
    fun initialize(contact: ParsedContact) {
        _parsedContact.value = contact

        val fullName = listOfNotNull(contact.givenName, contact.familyName)
            .joinToString(" ")
        _fullName.value = fullName

        _jobTitle.value = contact.jobTitle ?: ""
        _company.value = contact.organizationName ?: ""
        _email.value = contact.emailAddresses.firstOrNull()?.address ?: ""
        _phoneNumber.value = contact.phoneNumbers.firstOrNull()?.number ?: ""
        _website.value = contact.urls.firstOrNull()?.url ?: ""
        _notes.value = contact.note ?: ""
    }

    /**
     * Update field values
     */
    fun updateFullName(value: String) {
        _fullName.value = value
    }

    fun updateJobTitle(value: String) {
        _jobTitle.value = value
    }

    fun updateCompany(value: String) {
        _company.value = value
    }

    fun updateEmail(value: String) {
        _email.value = value
    }

    fun updatePhoneNumber(value: String) {
        _phoneNumber.value = value
    }

    fun updateWebsite(value: String) {
        _website.value = value
    }

    fun updateNotes(value: String) {
        _notes.value = value
    }

    /**
     * Save business card to database
     */
    fun saveCard() {
        viewModelScope.launch {
            _uiState.value = PreviewUiState.Saving

            try {
                val card = BusinessCard(
                    fullName = _fullName.value,
                    jobTitle = _jobTitle.value.takeIf { it.isNotBlank() },
                    company = _company.value.takeIf { it.isNotBlank() },
                    email = _email.value.takeIf { it.isNotBlank() },
                    phoneNumber = _phoneNumber.value.takeIf { it.isNotBlank() },
                    website = _website.value.takeIf { it.isNotBlank() },
                    notes = _notes.value.takeIf { it.isNotBlank() },
                    rawText = _parsedContact.value?.rawText ?: "",
                    dateScanned = Date(),
                    dateModified = Date()
                )

                repository.insertCard(card)
                _uiState.value = PreviewUiState.Success(card)
            } catch (e: Exception) {
                _uiState.value = PreviewUiState.Error(
                    e.message ?: "Failed to save card"
                )
            }
        }
    }

    /**
     * Save card and add to contacts
     */
    fun saveCardAndAddToContacts() {
        viewModelScope.launch {
            _uiState.value = PreviewUiState.Saving

            try {
                val card = BusinessCard(
                    fullName = _fullName.value,
                    jobTitle = _jobTitle.value.takeIf { it.isNotBlank() },
                    company = _company.value.takeIf { it.isNotBlank() },
                    email = _email.value.takeIf { it.isNotBlank() },
                    phoneNumber = _phoneNumber.value.takeIf { it.isNotBlank() },
                    website = _website.value.takeIf { it.isNotBlank() },
                    notes = _notes.value.takeIf { it.isNotBlank() },
                    rawText = _parsedContact.value?.rawText ?: "",
                    dateScanned = Date(),
                    dateModified = Date(),
                    savedToContacts = true
                )

                // Save to database
                repository.insertCard(card)

                // Add to Android Contacts
                if (permissionManager.hasContactsPermission()) {
                    contactsService.saveBusinessCardToContacts(card)
                }

                _uiState.value = PreviewUiState.Success(card)
            } catch (e: Exception) {
                _uiState.value = PreviewUiState.Error(
                    e.message ?: "Failed to save card"
                )
            }
        }
    }

    /**
     * Validate form
     */
    fun isValid(): Boolean {
        return _fullName.value.isNotBlank() &&
                (_email.value.isNotBlank() || _phoneNumber.value.isNotBlank())
    }
}

/**
 * UI states for Preview screen
 */
sealed class PreviewUiState {
    data object Editing : PreviewUiState()
    data object Saving : PreviewUiState()
    data class Success(val card: BusinessCard) : PreviewUiState()
    data class Error(val message: String) : PreviewUiState()
}
